src/librustc_mir/transform/validate.rs

```diff
@@ -38,6 +38,49 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             &format!("broken MIR in {:?} ({}): {}", self.def_id, self.when, msg.as_ref()),
         );
     }
+
+    /// Check if src can be assigned into dest.
+    /// This is not precise, it will accept some incorrect assignments.
+    fn mir_assign_valid_types(&self, src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+        if src == dest {
+            // Equal types, all is good.
+            return true;
+        }
+        // Normalize projections and things like that.
+        let src = self.tcx.normalize_erasing_regions(self.param_env, src);
+        let dest = self.tcx.normalize_erasing_regions(self.param_env, dest);
+        // It's worth checking equality again.
+        if src == dest {
+            return true;
+        }
+
+        // Type-changing assignments can happen for (at least) two reasons:
+        // 1. `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
+        // 2. Subtyping is used. While all normal lifetimes are erased, higher-ranked types
+        //    with their late-bound lifetimes are still around and can lead to type differences.
+        // Normalize both of them away.
+        // Also see the related but slightly different post-monomorphization method in `interpret/eval_context.rs`.
+        let normalize = |ty: Ty<'tcx>| {
+            ty.fold_with(&mut BottomUpFolder {
```

Comment:
Maybe using a `TypeRelation` feels more natural here.

Reply:
Maybe it does? I have never used that API.^^

Reply:
You can look at `SimpleEqRelation` for an example that does roughly what needs to be done here.
`tys` should check for equality before calling `super_relate_tys`, `regions` should unconditionally return `Some(a)`. `binders` should just skip the binders.

Reply:
@matthewjasper  I guess my question is, what is the benefit of that? And should [the related Miri check](https://github.com/rust-lang/rust/blob/033013cab3a861224fd55f494c8be1cb0349eb49/src/librustc_mir/interpret/eval_context.rs#L221) also be adjusted in your opinion?

Reply:
It is not clear to me if you say I should change this to land the PR, or if this is optional and could/should be done as a follow-up.

Reply:
It could be faster than folding because it doesn't create any new types. They probably should both be changed.
This PR can be landed without that change though. I think that the `&mut` to `&` transmute issue is fixed now, so the `mk_imm_ref` call below can be removed either way.

Reply:
> I think that the &mut to & transmute issue is fixed now, so the mk_imm_ref call below can be removed either way.

When was that fixed? A few weeks ago it was certainly still present.

Reply:
Also I wasn't even aware it was considered an issue, it seemed fine to do that as long as every MIR consumer is aware of this.

Reply:
#72093 and #72820.

Reply:
> tys should check for equality before calling super_relate_tys

What should it return? Seems like it is expected to return a type so I guess when both are equal I'll just return one of them.
(`tys` does not have a doc comment in the trait unfortunately.)

Reply:
For equality/subtyping relations `Ok(a)` is generally what's returned on success.